### PR TITLE
customiser for alpaka patatrack pixel seeding for phase2

### PR DESCRIPTION
**Pull request by lguzzi**
#### PR description:

this PR adds alpaka patatrack pixel tracking in the simplified menu using the [common alpaka process modifier](https://github.com/cms-sw/cmssw/blob/master/Configuration/ProcessModifiers/python/alpaka_cff.py).
The hltPhase2PixelTracks are redefined to be produced by CAHitNtupletAlpakaPhase2. Full tracks are produced merging initialStep (building from alpaka patatrack) + highPtTripletStep (building as in legacy). 

#### PR 
Alpaka tracking is enabled running the HLT step with --procModifiers alpaka:

```
cmsDriver.py step2  -s L1P2GT,HLT:75e33 --conditions auto:phase2_realistic_T33 --datatier GEN-SIM-DIGI-RAW -n 100 --eventcontent FEVTDEBUGHLT --geometry Extended2026D110 --era Phase2C17I13M9 --filein  file:step2_DIGIL1.root  --fileout file:step2_HLT.root  --nThreads 5 --procModifiers alpaka
```
(previous steps can be copied from ```runTheMatrix.py -l 29634.0 -t 4 -j 10 --ibeos --dryRun```)

Tracking performance are reported [here](https://lguzzi.web.cern.ch/lguzzi/Tracking/Phase2/alpaka/customiser/singleiter_vs_twoiters/generalTracks) for general tracks and [here](https://lguzzi.web.cern.ch/lguzzi/Tracking/Phase2/alpaka/customiser/singleiter_vs_twoiters/hltPhase2PixelTracks/) for pixel tracks.
When moving from the old customiser function to process modifiers I checked that the two give [identical results](https://lguzzi.web.cern.ch/lguzzi/Tracking/Phase2/alpaka/procModifier/xcheck/plots_hlt_hltGeneral/) (when disabling HCAL alpaka reconstruction).

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

not a backport

### CODE REVIEW

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+                backend = cms.untracked.string('')
+        )
+    )
+    process.hltOnlineBeamSpotDevice = cms.EDProducer('BeamSpotDeviceProducer@alpaka',
+            src = cms.InputTag('hltOnlineBeamSpot'),
+            alpaka = cms.untracked.PSet(
+                backend = cms.untracked.string('')
+        )
+    )
+    process.siPixelClustersSoA = cms.EDProducer('SiPixelPhase2DigiToCluster@alpaka',
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py:35]: Modules run at HLT start their label with ```hlt``` - this should now also be done for Phase-2.

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+        +process.siPixelClusters
+        +process.siPixelClusterShapeCache  # should we disable this? Still needed by tracker muons
+        #+process.siPixelDigis             # not needed when copying digis from sim
+        +process.siPixelRecHitsSoA
+        +process.siPixelRecHits
+    )
+    process.HLTDoLocalStripSequence = cms.Sequence(
+        process.siPhase2Clusters
+    )
+    process.itLocalRecoSequence = cms.Sequence(
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py:78]: Seuqences run at HLT start their label with ```HLT``` - please do so also for Phase-2.

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+def AlpakaHLTTrackingPath(process):
+    ''' Create a tracking-only path
+    '''
+    process.HLT_AlpakaTrackingPath = cms.Path()
+    process.HLT_AlpakaTrackingPath.insert(item=process.HLTBeginSequence         , index=len(process.HLT_AlpakaTrackingPath.moduleNames()))
+    process.HLT_AlpakaTrackingPath.insert(item=process.HLTTrackingV61Sequence   , index=len(process.HLT_AlpakaTrackingPath.moduleNames()))
+    process.HLT_AlpakaTrackingPath.insert(item=process.HLTEndSequence           , index=len(process.HLT_AlpakaTrackingPath.moduleNames()))
+
+    process.FEVTDEBUGHLTEventContent.outputCommands.append('keep *_siPixelClusters_*_reHLT')
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py:189]: Why is this specific processName ```reHLT``` hardwired here and below?

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    process.FEVTDEBUGHLTEventContent.outputCommands.append('keep *_siPixelRecHits_*_reHLT')
+    process.FEVTDEBUGHLTEventContent.outputCommands.append('keep *_hltPhase2PixelTracks_*_reHLT')
+    process.FEVTDEBUGHLTEventContent.outputCommands.append('keep *_generalTracks_*_reHLT')
+    process.FEVTDEBUGHLTEventContent.outputCommands.append('drop IntermediateHitDoublets_hltElePixelHitDoubletsL1Seeded__reHLT')            # remove warning
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py:193]: Syntax is incorrect - I suppose ```L1Seeded_*_reHLT``` is required here and below?

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        +process.siPixelClusterShapeCache  # should we disable this? Still needed by tracker muons
+        #+process.siPixelDigis             # not needed when copying digis from sim
+        +process.siPixelRecHitsSoA
+        +process.siPixelRecHits
+    )
+    process.HLTDoLocalStripSequence = cms.Sequence(
+        process.siPhase2Clusters
+    )
+    process.itLocalRecoSequence = cms.Sequence(
+         process.HLTDoLocalStripSequence
```
Comment (mmusich) [inline HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py:79]: Why do you call this Strips for phase2? Something containing OT would be more appropriate

Comment (mmusich) [reply HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py:79]: Also I find it confusing you call an OT local reco sequence within `itLocalRecoSequence` (again the naming convention is not respected).

**Comment by mmusich** [comment]
assign heterogeneous

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py
Comment (fwyzard) [inline HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py:1]: I wouldn't put `141X` in the file name, as it is likely to last into future releases as well.

**Comment by fwyzard** [comment]
@rovere @VourMa @SohamBhattacharya FYI

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py (hunk 4 above)
Comment (lguzzi) [reply HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py]: these were leftovers, removed them in 2147f3d41dc5f1f27929683ee5f16bda0dfde052

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py (hunk 3 above)
Comment (lguzzi) [reply HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py]: these were leftovers, removed them in 2147f3d41dc5f1f27929683ee5f16bda0dfde052

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py (hunk 1 above)
Comment (lguzzi) [reply HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py]: I changed the name only for the new collections added by the customiser in beb3cb2d8258f29fc2fd33ab4c12b11964e00872
There are several modules and sequences in the phase2 menu which do not follow the naming convention. This problem is known, it will be fixed in a future PR

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py (hunk 2 above)
Comment (lguzzi) [reply HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py]: There are several modules and sequences in the phase2 menu which do not follow the naming convention. This problem is known, it will be fixed in a future PR. This sequence in particular is already defined in https://github.com/cms-sw/cmssw/blob/0cbc4b00b1664496105480e9f561a29ba7c9a68b/HLTrigger/Configuration/python/HLT_75e33/sequences/HLTDoLocalPixelSequence_cfi.py#L4 and I removed it from the customiser, since the default definition is ok (208d97d6f5c4dc677b0ce3d11e5d5db357e97785).

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py (hunk 5 above)
Comment (lguzzi) [reply HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py]: this sequence is already defined in https://github.com/cms-sw/cmssw/blob/0cbc4b00b1664496105480e9f561a29ba7c9a68b/HLTrigger/Configuration/python/HLT_75e33/sequences/HLTDoLocalPixelSequence_cfi.py . Its redefinition here is actually not needed, so I removed it from the customiser in 208d97d6f5c4dc677b0ce3d11e5d5db357e97785. The problem of phase2 naming convention will be addressed in a future PR

Comment (lguzzi) [reply HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py]: the same is true for https://github.com/cms-sw/cmssw/blob/0cbc4b00b1664496105480e9f561a29ba7c9a68b/HLTrigger/Configuration/python/HLT_75e33/sequences/itLocalRecoSequence_cfi.py, which calls siPhase2Clusters as well. I did not change the structure of already existing sequences because they are called by other paths.

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py
Comment (lguzzi) [reply HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py:1]: done in 517e816bf4bb0a838d83af754eca52e04c8a061c

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py (hunk 2 above)
Comment (mmusich) [reply HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py]: Mmmh, as long as the new instances are used only in the customization function why do you need to keep the naming used elsewhere? Aren't you replacing the sequence entirely?

**Comment by fwyzard** [comment]
In general, keeping the customisation close to the module or sequence that it affects makes it easier to see what impact it has.

Modifiers are a way to do that.

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py (hunk 2 above)
Comment (lguzzi) [reply HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py]: the idea was to allow other POGs to use the customiser with minimal effort. Some paths include this sequence (for example https://github.com/cms-sw/cmssw/blob/0cbc4b00b1664496105480e9f561a29ba7c9a68b/HLTrigger/Configuration/python/HLT_75e33/paths/HLT_Mu37_Mu27_FromL1TkMuon_cfi.py#L19, but it is true for other sequences as well), and using a new name for the sequence would not propagate the changes done in the customiser to these paths

**Comment by fwyzard** [comment]
It's also, for better or worse, the approach generally recommended for CMSSW configurations.

For the Run-3 HLT menu it _cannot_ be applied because we keep the configuration as a single large file that we dump from ConfDB.

For the Run 4+ HLT menu we currently _can_ do it, so I would be inclined to follow the more widespread approach.

**Comment by mmusich** [comment]
> For the Run 4+ HLT menu we currently can do it, so I would be inclined to follow the more widespread approach.

Since we have the precedent of the hgcal @ alpaka, I have no reason to object now, but before making this the general recommendation for developers I'd like to have the item more widely discussed.

File: HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py (hunk 5 above)
Comment (mmusich) [reply HLTrigger/Configuration/python/customizer_AlpakaPixelTracking_Phase2_141X.py]: > The problem of phase2 naming convention will be addressed in a future PR

@lguzzi please take note of https://github.com/cms-sw/cmssw/pull/45500#issuecomment-2242744894 (the renaming was kindly taken care of by  @AuroraPerego). Please cross-check if you have comments.

**Comment by slava77** [comment]
type tracking

**Comment by mmusich** [comment]
-1
   * as discussed at the the [Jul 2023 HLT upgrade meeting](https://indico.cern.ch/event/1439974/), this needs to be reworked as requested at https://github.com/cms-sw/cmssw/pull/45508#issuecomment-2239072422.

**Comment by antoniovilela** [comment]
ping (to make bot change milestone)

File: HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPPixelCPEFastParams_cfi.py [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import FWCore.ParameterSet.Config as cms
+
+hltESPPixelCPEFastParamsPhase2 = cms.ESProducer('PixelCPEFastParamsESProducerAlpakaPhase2@alpaka', 
+    ComponentName = cms.string("PixelCPEFastParamsPhase2"),
+    appendToDataLabel = cms.string(''),
+    alpaka = cms.untracked.PSet(backend = cms.untracked.string(''))
+)
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPPixelCPEFastParams_cfi.py:7]:
```suggestion
)
```

here and elsewhere, please keep the newlines at the end of files.

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2PixelTracksSoA_cfi.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        # https://github.com/cms-sw/cmssw/blob/CMSSW_14_1_X/RecoTracker/PixelSeeding/plugins/CAHitNtupletGeneratorOnGPU.cc#L253-L257
+        #chi2MaxPt = cms.double(10),
+        #chi2Coeff = cms.vdouble(0.9, 1.8),
+        #chi2Scale = cms.double(8),
+        #tripletMinPt = cms.double(0.5),
+        #tripletMaxTip = cms.double(0.3),
+        #tripletMaxZip = cms.double(12),
+        #quadrupletMinPt = cms.double(0.3),
+        #quadrupletMaxTip = cms.double(0.5),
+        #quadrupletMaxZip = cms.double(12)
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2PixelTracksSoA_cfi.py:46]: what's the point of keeping these commented lines?

File: HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPSiPixelCablingSoA_cfi.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+def _addProcessSiPixelCablingAlpaka(process):
+    process.hltESPSiPixelCablingSoA = cms.ESProducer('SiPixelCablingSoAESProducer@alpaka', 
+        CablingMapLabel = cms.string(''),
+        UseQualityInfo = cms.bool(False),
+        appendToDataLabel = cms.string(''),
+        alpaka = cms.untracked.PSet(backend = cms.untracked.string(''))
+    )
+
+from Configuration.ProcessModifiers.alpaka_cff import alpaka
+modifyConfigurationForAlpakaSiPixelCabling_ = alpaka.makeProcessModifier(_addProcessSiPixelCablingAlpaka)
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPSiPixelCablingSoA_cfi.py:12]: missing newline.

File: HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPSiPixelGainCalibrationForHLTSoA_cfi.py [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import FWCore.ParameterSet.Config as cms
+
+def _addProcessSiPixelGainCalibrationAlpaka(process):
+    process.hltESPSiPixelGainCalibrationForHLTSoA = cms.ESProducer('SiPixelGainCalibrationForHLTSoAESProducer@alpaka',
+        appendToDataLabel = cms.string(''),
+        alpaka = cms.untracked.PSet(backend = cms.untracked.string(''))
+    )
+
+from Configuration.ProcessModifiers.alpaka_cff import alpaka
+modifyConfigurationForAlpakaSiPixelGainCalibration_ = alpaka.makeProcessModifier(_addProcessSiPixelGainCalibrationAlpaka)
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPSiPixelGainCalibrationForHLTSoA_cfi.py:10]: missing newline.

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2OnlineBeamSpotDevice_cfi.py [hunk 10]
```diff
@@ -0,0 +1,6 @@
+import FWCore.ParameterSet.Config as cms
+
+hltPhase2OnlineBeamSpotDevice = cms.EDProducer('BeamSpotDeviceProducer@alpaka',
+    src = cms.InputTag('hltOnlineBeamSpot'),
+    alpaka = cms.untracked.PSet(backend = cms.untracked.string(''))
+)
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2OnlineBeamSpotDevice_cfi.py:6]: missing newline.

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2SiPixelClustersSoA_cfi.py [hunk 11]
```diff
@@ -0,0 +1,8 @@
+import FWCore.ParameterSet.Config as cms
+
+hltPhase2SiPixelClustersSoA = cms.EDProducer(
+    "SiPixelPhase2DigiToCluster@alpaka",
+    alpaka = cms.untracked.PSet(
+        backend = cms.untracked.string('')
+    )
+)
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2SiPixelClustersSoA_cfi.py:8]: missing newline.

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2SiPixelRecHitsSoA_cfi.py [hunk 12]
```diff
@@ -0,0 +1,10 @@
+import FWCore.ParameterSet.Config as cms
+
+hltPhase2SiPixelRecHitsSoA = cms.EDProducer('SiPixelRecHitAlpakaPhase2@alpaka',
+    beamSpot = cms.InputTag('hltPhase2OnlineBeamSpotDevice'),
+    src = cms.InputTag('hltPhase2SiPixelClustersSoA'),
+    CPE = cms.string('PixelCPEFastParamsPhase2'),
+    mightGet = cms.optional.untracked.vstring,
+    # autoselect the alpaka backend
+    alpaka = cms.untracked.PSet(backend = cms.untracked.string(''))
+)
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2SiPixelRecHitsSoA_cfi.py:10]: missing newline.

File: HLTrigger/Configuration/python/HLT_75e33/sequences/HLTDoLocalPixelSequence_cfi.py [hunk 13]
```diff
@@ -4,3 +4,18 @@
...
 HLTDoLocalPixelSequence = cms.Sequence(hltSiPixelClusters+hltSiPixelRecHits)
+
+from ..modules.hltPhase2SiPixelClustersSoA_cfi import hltPhase2SiPixelClustersSoA
+from ..modules.hltPhase2SiPixelRecHitsSoA_cfi  import hltPhase2SiPixelRecHitsSoA
+from ..modules.hltSiPixelClusterShapeCache_cfi import hltSiPixelClusterShapeCache
+_HLTDoLocalPixelSequence = cms.Sequence(
+     hltPhase2SiPixelClustersSoA
+    +hltSiPixelClusters
+    +hltSiPixelClusterShapeCache  # should we disable this? Still needed by tracker muons
+    #+siPixelDigis             # not needed when copying digis from sim
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/sequences/HLTDoLocalPixelSequence_cfi.py:15]:
```suggestion
```

then remove.

File: HLTrigger/Configuration/python/HLT_75e33/sequences/HLTItLocalRecoSequence_cfi.py [hunk 14]
```diff
@@ -6,3 +6,10 @@
 from ..modules.hltSiPixelRecHits_cfi import *
 
 HLTItLocalRecoSequence = cms.Sequence(hltSiPhase2Clusters+hltSiPixelClusters+hltSiPixelClusterShapeCache+hltSiPixelRecHits)
+
+from ..sequences.HLTDoLocalPixelSequence_cfi import *
+from ..sequences.HLTDoLocalStripSequence_cfi import *
+_HLTItLocalRecoSequence = cms.Sequence(HLTDoLocalPixelSequence+HLTDoLocalStripSequence)
+
+from Configuration.ProcessModifiers.alpaka_cff import alpaka
+alpaka.toReplaceWith(HLTItLocalRecoSequence, _HLTItLocalRecoSequence)
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLT_75e33/sequences/HLTItLocalRecoSequence_cfi.py:15]: missing newline.

**Comment by mmusich** [comment]
enable gpu

**Comment by mmusich** [comment]
test parameters:
   * workflows = 29634.402
   * relvals_opt = --what upgrade

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2PixelTracksSoA_cfi.py (hunk 7 above)
Comment (lguzzi) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2PixelTracksSoA_cfi.py]: I removed them in the last commit

**Comment by mmusich** [comment]
@cmsbuild, please test

File: RecoLocalTracker/SiPixelRecHits/python/PixelCPEGeneric_cfi.py [hunk 15]
```diff
@@ -54,3 +54,7 @@
 (phase2_tracker & (phase2_squarePixels | phase2_3DPixels)).toModify(PixelCPEGenericESProducer,
                                                                     NoTemplateErrorsWhenNoTrkAngles = False # use genErrors in the seeding step (when no track angles are available)
                                                                     )
+from Configuration.ProcessModifiers.alpaka_cff import alpaka
```
Comment (mmusich) [inline RecoLocalTracker/SiPixelRecHits/python/PixelCPEGeneric_cfi.py:57]: @lguzzi 

this is misplaced in this file as it overwrites the settings from the reconstruction with the HLT menu one, see the log of [wf  12434.402](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e4bd8b/41250/runTheMatrixGPU-results/12434.402_TTbar_14TeV+2023_Patatrack_PixelOnlyAlpaka/step3_TTbar_14TeV+2023_Patatrack_PixelOnlyAlpaka.log)

```
----- Begin Fatal Exception 03-Sep-2024 14:58:48 CEST-----------------------
An exception of category 'EventSetupConflict' occurred while
   [0] Calling beginJob
Exception Message:
two EventSetup Producers want to deliver type="PixelCPEFastParamsHostPhase2" label="PixelCPEFastParamsPhase2"
 from record PixelCPEFastParamsRecord. The two providers are 
1) type="PixelCPEFastParamsESProducerAlpakaPhase2@alpaka" label="hltESPPixelCPEFastParamsPhase2"
2) type="PixelCPEFastParamsESProducerAlpakaPhase2@alpaka" label="pixelCPEFastParamsESProducerAlpakaPhase2"
Please either
   remove one of these Producers
   or find a way of configuring one of them so it does not deliver this data
   or use an es_prefer statement in the configuration to choose one.
----- End Fatal Exception -------------------------------------------------
```

please make this change self-contained in the phase-2 HLT menu.

Comment (lguzzi) [reply RecoLocalTracker/SiPixelRecHits/python/PixelCPEGeneric_cfi.py]: done in 78974e0384070734cc0e58c72f1b636004558fa2

**Comment by mmusich** [comment]
@cmsbuild, please test

File: HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPPixelCPEFastParams_cfi.py [hunk 16]
```diff
@@ -0,0 +1,13 @@
+import FWCore.ParameterSet.Config as cms
+
+
+def _addProcessPixelCPEFastParamsPhase2(process):
+    process.hltESPPixelCPEFastParamsPhase2 = cms.ESProducer('PixelCPEFastParamsESProducerAlpakaPhase2@alpaka', 
+    ComponentName = cms.string("PixelCPEFastParamsPhase2"),
+    appendToDataLabel = cms.string(''),
+    alpaka = cms.untracked.PSet(backend = cms.untracked.string(''))
+)
+
```
Comment (fwyzard) [inline HLTrigger/Configuration/python/HLT_75e33/eventsetup/hltESPPixelCPEFastParams_cfi.py:10]: pedantic suggestion:
```suggestion
def _addProcessPixelCPEFastParamsPhase2(process):
    process.hltESPPixelCPEFastParamsPhase2 = cms.ESProducer('PixelCPEFastParamsESProducerAlpakaPhase2@alpaka', 
        ComponentName = cms.string("PixelCPEFastParamsPhase2"),
        appendToDataLabel = cms.string(''),
        alpaka = cms.untracked.PSet(backend = cms.untracked.string('')
    )
)
```

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2SiPixelClustersSoA_cfi.py [hunk 17]
```diff
@@ -0,0 +1,8 @@
+import FWCore.ParameterSet.Config as cms
+
+hltPhase2SiPixelClustersSoA = cms.EDProducer(
+    "SiPixelPhase2DigiToCluster@alpaka",
+    alpaka = cms.untracked.PSet(
+        backend = cms.untracked.string('')
+    )
```
Comment (fwyzard) [inline HLTrigger/Configuration/python/HLT_75e33/modules/hltPhase2SiPixelClustersSoA_cfi.py:7]: Pedantic suggestion, for consistency with the other modules:
```suggestion
hltPhase2SiPixelClustersSoA = cms.EDProducer("SiPixelPhase2DigiToCluster@alpaka",
    # autoselect the alpaka backend
    alpaka = cms.untracked.PSet(backend = cms.untracked.string(''))
```

**Comment by mmusich** [comment]
@lguzzi before restarting the tests, would you mind to squash the commits to 1?

**Comment by lguzzi** [comment]
> @lguzzi before restarting the tests, would you mind to squash the commits to 1?

done

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
> done

thanks. @lguzzi as of now, the PR description is out of date as it still refers to customization functions.
Can you please update it and include a cmsDriver command on how to test it standalone?

**Comment by lguzzi** [comment]
> > done
> 
> thanks. @lguzzi as of now, the PR description is out of date as it still refers to customization functions. Can you please update it and include a cmsDriver command on how to test it standalone?

done

**Comment by mmusich** [comment]
> -1

OK, the relval failure looks spurious.
About the changes:
   * [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-03-1100+e4bd8b/64321/triggerResults): found differences in 1 / 44 workflows => this happens in `29634.402_TTbar_14TeV+2026D110_Patatrack_PixelOnlyAlpaka` and looks just a re-ordering of the of the trigger bits (no changes in accepted or lost)
   *  You potentially [added 27 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-03-1100+e4bd8b/64321/validateJR/logRootQA-events.log) to the logs seems to be due to
   
   ```
   4 Lines only in /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-e4bd8b/24834.911_TTbar_14TeV+2026D98_DD4hep/step2_TTbar_14TeV+2026D98_DD4hep.log
   %MSG-i AlpakaService:  (NoModuleName) DATETIME CEST pre-events
   AlpakaServiceSerialSync succesfully initialised.
   Found 1 device:
   - Intel(R) Xeon(R) CPU E5-2683 v4 @ 2.10GHz
   ```
   
   which probably comes from https://github.com/cms-sw/cmssw/blob/c8155d4d580463a385aea1ba55299a9a33079432/HLTrigger/Configuration/python/HLT_75e33_cff.py#L6
   * changes in DQM and RECO products (e.g. [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-03-1100+e4bd8b/64321/validateJR.html)) are concentrated in `12434.7` and `12834.7` (somewhat unstable full-blown mkFit workflows), see also: https://github.com/cms-sw/cmssw/issues/39803

**Comment by mmusich** [comment]
I would re-trigger the tests to get a cleaner comparison. Let's wait though if @fwyzard has more comments on the structure of the change.

**Comment by mmusich** [comment]
test parameters:
   * enable = gpu
   * workflows_gpu = 29634.402, 31834.492
   * workflow_opts = -w upgrade
   * workflow_opts_gpu = -w upgrade

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
In the last round of tests in which we got also the DQM comparisons for e.g. [31834.492 overlay GUI](https://cmsweb.cern.ch/dqm/dev/start?runnr=1;dataset%3D/RelVal_wf31834_492_pr/CMSSW_14_2_X-PRcmssw_45508-64344/DQMIO;referenceshow%3Dall;referencenorm=False;referenceobj1%3Dother::/RelVal_wf31834_492_base/CMSSW_14_2_X-PRcmssw_45508-64344/DQMIO::;sampletype%3Doffline_relval;workspace%3DEverything;).
As expected there are rather large improvements:
   * HLT pixel tracks: https://tinyurl.com/2bv6gyx3
   * HLT pixel vertices: https://tinyurl.com/28zehson

**Comment by mmusich** [comment]
+hlt
   * see https://github.com/cms-sw/cmssw/pull/45508#issuecomment-2330933424

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltSiPixelClusters_cfi.py [hunk 18]
```diff
@@ -22,3 +22,13 @@
...
+
+from Configuration.ProcessModifiers.alpaka_cff import alpaka
+_hltSiPixelClusters = cms.EDProducer('SiPixelDigisClustersFromSoAAlpakaPhase2',
+    src = cms.InputTag('hltPhase2SiPixelClustersSoA'),
+    clusterThreshold_layer1 = cms.int32(4000),
+    clusterThreshold_otherLayers = cms.int32(4000),
+    produceDigis = cms.bool(False),
+    storeDigis = cms.bool(False)
+)
+alpaka.toReplaceWith(hltSiPixelClusters, _hltSiPixelClusters.clone())
```
Comment (rovere) [inline HLTrigger/Configuration/python/HLT_75e33/modules/hltSiPixelClusters_cfi.py:34]: why do we need the additional `.clone()`?

**Comment by rovere** [comment]
would it make sense to also update the `timing` version of the HLT Phase2 menu?

**Comment by rovere** [comment]
enable hlt_p2_timing

**Comment by rovere** [comment]
please test

**Comment by rovere** [comment]
> would it make sense to also update the `timing` version of the HLT Phase2 menu?

Let me rephrase: it would be nice to verify if the `timing` variant of the HLT menu could work properly with the `alpaka` procModifier enabled.

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltSiPixelClusters_cfi.py (hunk 18 above)
Comment (lguzzi) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltSiPixelClusters_cfi.py:34]: it is a leftover, probably not needed. Should I remove it?

Comment (mmusich) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltSiPixelClusters_cfi.py:34]: Yes, please

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
+hlt
   * re-sign

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by mandrenguyen** [comment]
+1
